source/common/buffer/buffer_impl.cc

```diff
@@ -43,71 +56,213 @@ void OwnedImpl::add(const Instance& data) {
+  uint64_t num_slices_committed = 0;
+  while (num_slices_committed < num_iovecs) {
+    if (slices_[slice_index]->commit(iovecs[num_slices_committed])) {
+      length_ += iovecs[num_slices_committed].len_;
+      num_slices_committed++;
+    }
+    slice_index++;
+    if (slice_index == static_cast<ssize_t>(slices_.size())) {
+      break;
+    }
+  }
 
-  evbuffer_ptr start_ptr;
-  int rc = evbuffer_ptr_set(buffer_.get(), &start_ptr, start, EVBUFFER_PTR_SET);
-  ASSERT(rc != -1);
+  ASSERT(num_slices_committed > 0);
+}
 
-  ev_ssize_t copied = evbuffer_copyout_from(buffer_.get(), &start_ptr, data, size);
-  ASSERT(static_cast<uint64_t>(copied) == size);
+void OwnedImpl::copyOut(size_t start, uint64_t size, void* data) const {
+  uint8_t* dest = static_cast<uint8_t*>(data);
+  for (size_t slice_index = 0; slice_index < slices_.size(); slice_index++) {
+    const auto& slice = slices_[slice_index];
+    if (size == 0) {
+      break;
+    }
+    uint64_t data_size = slice->dataSize();
+    if (data_size <= start) {
+      start -= data_size;
+      continue;
+    }
+    uint64_t copy_size = std::min(size, data_size - start);
+    memcpy(dest, static_cast<const uint8_t*>(slice->data()) + start, copy_size);
+    size -= copy_size;
+    dest += copy_size;
+    start = 0;
+  }
+  ASSERT(size == 0);
 }
 
 void OwnedImpl::drain(uint64_t size) {
-  ASSERT(size <= length());
-  int rc = evbuffer_drain(buffer_.get(), size);
-  ASSERT(rc == 0);
+  while (size != 0) {
+    if (slices_.empty()) {
+      break;
+    }
+    uint64_t slice_size = slices_.front()->dataSize();
+    if (slice_size <= size) {
+      slices_.pop_front();
+      length_ -= slice_size;
+      size -= slice_size;
+    } else {
+      slices_.front()->drain(size);
+      length_ -= size;
+      size = 0;
+    }
+  }
 }
 
 uint64_t OwnedImpl::getRawSlices(RawSlice* out, uint64_t out_size) const {
-  return evbuffer_peek(buffer_.get(), -1, nullptr, reinterpret_cast<evbuffer_iovec*>(out),
-                       out_size);
+  uint64_t i = 0;
+  for (size_t slice_index = 0; slice_index < slices_.size(); slice_index++) {
+    const auto& slice = slices_[slice_index];
+    if (slice->dataSize() == 0) {
+      continue;
+    }
+    if (i < out_size) {
+      // The cumbersome cast here allows RawSlice::mem_ to remain immutable by everything
+      // outside OwnedImpl. The comments accompanying the RawSlice declaration in
+      // include/envoy/buffer/buffer.h provide more context on why the immutability
+      // is important.
+      *(const_cast<void**>(&(out[i].mem_))) = slice->data();
+      out[i].len_ = slice->dataSize();
+    }
+    i++;
+  }
+  return i;
 }
 
-uint64_t OwnedImpl::length() const { return evbuffer_get_length(buffer_.get()); }
+uint64_t OwnedImpl::length() const {
+  // When running in debug mode, verify that the precomputed length matches the sum
+  // of the lengths of the slices.
+  // NOLINTNEXTLINE(bugprone-assert-side-effect)
```

Comment:
Why did you need this? Your ASSERT shouldn't have side effect, it might be possibly due to the for loop isn't const range loop.